// obsidian-mcp-server/build/types.js
export const DEFAULT_OBSIDIAN_CONFIG = {
    protocol: "https", // HTTPS required by default in Obsidian REST API plugin
    host: "127.0.0.1",
    port: 27124
};
export const DEFAULT_RATE_LIMIT_CONFIG = {
    windowMs: 15 * 60 * 1000, // 15 minutes
    maxRequests: 200
};
export class ObsidianError extends Error {
    details;
    errorCode;
    constructor(message, errorCode = 50000, // Default server error code
    details) {
        super(message);
        this.details = details;
        this.name = "ObsidianError";
        // Ensure 5-digit error code
        if (errorCode < 10000 || errorCode > 99999) {
            // Convert HTTP status codes to 5-digit codes
            // 4xx -> 4xxxx
            // 5xx -> 5xxxx
            this.errorCode = errorCode < 1000 ? errorCode * 100 : 50000;
        }
        else {
            this.errorCode = errorCode;
        }
    }
    // Convert to API error format
    toApiError() {
        return {
            errorCode: this.errorCode,
            message: this.message
        };
    }
}
